### PR TITLE
A lot of stuff

### PR DESCRIPTION
**Pull request by esjee**
The goal of this diff is to present the user with a view that allows them to track changes in an example.

The example page now shows the change (red if total nu. of queries increased, green if the total nu. of queries decreased). There's a link to a comparison between two results, and a result view.

The comparison lists a diff between two results, underlining queries that got added and crossing out queries that were removed. Unchanged queries are presented plainly.

The result view simply shows all queries for a given result.

### CODE REVIEW

File: app/controllers/results_controller.rb [hunk 1]
```diff
@@ -5,6 +5,16 @@ def create
     render :ok, json: create_results
   end
 
+  def index
+    @results = Result.where(id: params[:id])
```
Comment (Willianvdv) [inline app/controllers/results_controller.rb]: Missing auth, we should limit the access to #create using the API key and #index using the current_user. Also fetch the results using the `project` instance

File: app/helpers/results_comper.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class ResultsComper
+  attr_reader :left
+  attr_reader :right
```
Comment (Willianvdv) [inline app/helpers/results_comper.rb:3]: can be private

File: app/helpers/results_viewer.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class ResultsViewer
+  attr_reader :results
```
Comment (Willianvdv) [inline app/helpers/results_viewer.rb]: can be private

File: app/helpers/results_viewer.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  def initialize(results)
+    @results = results
+  end
+
+  def each_with_maybe_previous
```
Comment (Willianvdv) [inline app/helpers/results_viewer.rb:6]: call me maybe pattern

File: app/helpers/results_viewer.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    previous = nil
+    results.each do |current|
+      if previous
+        query_change = previous.queries_count - current.queries_count
+      else
+        query_change = 0
+      end
```
Comment (Willianvdv) [inline app/helpers/results_viewer.rb:13]:
```
query_change = if previous
  previous.queries_count - current.queries_count
else
  query_change = 0
end
```

File: app/controllers/results_controller.rb [hunk 6]
```diff
@@ -5,6 +5,16 @@ def create
...
+  end
+
+  def compare_view
+    @result_left = Result.find(params[:result_id_left])
+    @result_right = Result.find(params[:result_id_right])
```
Comment (Willianvdv) [inline app/controllers/results_controller.rb]: ^ The comper has this data. you could use `@results_comper.left`

File: app/helpers/results_comper.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  def initialize(left, right)
+    @left = left
+    @right = right
+  end
+
+  def diffy
+    Diffy::Diff.new(
+      left.queries.map(&:statement).join("\n"),
+      right.queries.map(&:statement).join("\n"),
+    ).to_s(:html)
```
Comment (Willianvdv) [inline app/helpers/results_comper.rb:14]: potential xss

File: app/views/results/compare_view.html.erb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<h1>Comparison from <%= link_to(@result_left.id, "/results?id=#{@result_left.id}") %> to <%= link_to(@result_right.id, "/results?id=#{@result_right.id}") %></h1>
+
+<%= @results_comper.diffy.html_safe %>
```
Comment (Willianvdv) [inline app/views/results/compare_view.html.erb:3]: `<%= h @results_comper.diffy %>`

trust :all_the_things:

File: app/controllers/results_controller.rb (hunk 1 above)
Comment (esjee) [inline app/controllers/results_controller.rb]: > #create using the API key and #index using the 

In this PR?

> Also fetch the results using the project instance

Fetching using user instead (fetching a results shouldn't you require to specify the project)

File: app/helpers/results_comper.rb (hunk 2 above)
Comment (esjee) [inline app/helpers/results_comper.rb:3]: Instead removed "left_result", "right_result" from controller.

File: app/helpers/results_viewer.rb (hunk 4 above)
Comment (esjee) [inline app/helpers/results_viewer.rb:6]: Yeah, the name is silly. The previous value is sometimes `nil`; hence "maybe".

File: app/controllers/results_controller.rb (hunk 6 above)
Comment (esjee) [inline app/controllers/results_controller.rb]: Fixed

File: app/views/results/compare_view.html.erb (hunk 8 above)
Comment (esjee) [inline app/views/results/compare_view.html.erb:3]: `h` is an alias for `html_escape` (http://api.rubyonrails.org/classes/ERB/Util.html#method-c-h); //not// `html_safe`

**Pull request merged**